test/asset-swapper/utils/market_operations_utils/path_test.ts

```diff
@@ -0,0 +1,76 @@
+import * as chai from 'chai';
+import 'mocha';
+import {
+    BigNumber,
+    ERC20BridgeSource,
+    Fill,
+    FillQuoteTransformerOrderType,
+    MarketOperation,
+} from '../../../../src/asset-swapper';
+import { ONE_ETHER } from '../../../../src/asset-swapper/utils/market_operation_utils/constants';
+import { Path } from '../../../../src/asset-swapper/utils/market_operation_utils/path';
+import { chaiSetup } from '../chai_setup';
+
+chaiSetup.configure();
+const expect = chai.expect;
+
+describe('Path', () => {
+    describe('adjustedRate()', () => {
+        it('Returns the adjusted rate based on adjustedOutput and exchange proxy overhead', () => {
+            const path = Path.create(
+                MarketOperation.Sell,
+                [createFakeFill({ input: ONE_ETHER, adjustedOutput: ONE_ETHER.times(990) })],
+                ONE_ETHER,
+                {
+                    inputAmountPerEth: new BigNumber(1),
+                    outputAmountPerEth: new BigNumber(1000),
+                    exchangeProxyOverhead: () => ONE_ETHER.times(0.01), // 10 * 10e18 output amount
+                },
+            );
+
+            // 990 (adjusted output) - 10 (overhead)
+            expect(path.adjustedRate()).bignumber.eq(new BigNumber(990 - 10));
+        });
+
+        it('Returns the adjusted rate without interpolating penalty when sum of the input amounts is greater than the target input amount', () => {
+            const path = Path.create(
+                MarketOperation.Sell,
+                [
+                    createFakeFill({
+                        input: ONE_ETHER,
+                        adjustedOutput: ONE_ETHER.times(990),
+                    }),
+                    createFakeFill({
+                        input: ONE_ETHER,
+                        output: ONE_ETHER.times(1000),
```

Comment:
how come sometimes there is output but sometimes only adjustedOutput?

Reply:
`output` is only used when the sum of inputs is greater than the target. 

I suspect that this does not happen in prod though 😅 (I didn't want to change any behavior in this PR since it's a refactoring PR)